lib/net/wifi.toit

```diff
@@ -15,26 +16,55 @@ import ..esp32
 WIFI_CONNECT_TIMEOUT_  ::= Duration --s=10
 WIFI_DHCP_TIMEOUT_     ::= Duration --s=16
 
+wifi_/wifi.Wifi? := null
+wifi_connecting_/monitor.Latch? := null
+
 connect --ssid/string?=null --password/string="" -> net.Interface:
   if not ssid:
     config := image_config or {:}
     wifi_config := config.get "wifi" --if_absent=: {:}
     ssid = wifi_config.get "ssid"
     password = wifi_config.get "password"
 
-  wifi := wifi.Wifi
-  try:
-    wifi.set_ssid ssid password
-    with_timeout WIFI_CONNECT_TIMEOUT_: wifi.connect
-    with_timeout WIFI_DHCP_TIMEOUT_: wifi.get_ip
-    return WifiInterface_ wifi
-  finally: | is_exception _ |
-    if is_exception: wifi.close
+  if not wifi_:
+    if wifi_connecting_:
+      // If we're already connecting, we wait and see if that leads to
+      // an exception. If it doesn't throw, then we continue to mark
+      // ourselves as a user of the WiFi network.
+      exception := wifi_connecting_.get
+      if exception: throw exception
+    else:
+      // We use a latch to coordinate the initial creation of the WiFi
+      // connection. We always set it to a value when leaving this path
+      // and we always clear out the latch when we're done synchronizing.
+      wifi_connecting_ = monitor.Latch
+      wifi := wifi.Wifi
+      try:
+        wifi.set_ssid ssid password
+        with_timeout WIFI_CONNECT_TIMEOUT_: wifi.connect
+        with_timeout WIFI_DHCP_TIMEOUT_: wifi.get_ip
+        // Success: Register the WiFi connection, tell anyone who is waiting
+        // for it that the connection is ready to be used (no exception), and
+        // go on to mark ourselves as a user of the WiFi network.
+        wifi_ = wifi
+        wifi_connecting_.set null
+      finally: | is_exception exception |
+        if is_exception:
+          wifi.close
+          wifi_connecting_.set exception.value
+        wifi_connecting_ = null
+
+  interface := WifiInterface_
+  if ssid == wifi_.ssid_ and password == wifi_.password_: return interface
+  interface.close
```

Comment:
Why even allocate the interface if we are not going to use it?

Would be simpler to just:
```
if ssid == wifi_.ssid_ and password == wifi_.password_: return WifiInterface_
```

Reply:
The wifi_ isn't closed.